sync_diff_inspector/source/common/conn.go

```diff
@@ -15,31 +15,60 @@ package common
 
 import (
 	"database/sql"
-	"regexp"
+	"encoding/base64"
 
+	"github.com/go-sql-driver/mysql"
 	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	tmysql "github.com/pingcap/tidb/parser/mysql"
 )
 
-// CreateDB creates sql.DB used for select data
-func CreateDB(dsn string, num int) (db *sql.DB, err error) {
-	db, err = sql.Open("mysql", dsn)
+func tryConnectMySQL(cfg *mysql.Config) (*sql.DB, error) {
+	failpoint.Inject("MustMySQLPassword", func(val failpoint.Value) {
+		pwd := val.(string)
+		if cfg.Passwd != pwd {
+			failpoint.Return(nil, &mysql.MySQLError{Number: tmysql.ErrAccessDenied, Message: "access denied"})
+		}
+		failpoint.Return(nil, nil)
+	})
+	c, err := mysql.NewConnector(cfg)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	db := sql.OpenDB(c)
+	if err = db.Ping(); err != nil {
+		_ = db.Close()
+		return nil, errors.Trace(err)
+	}
+	return db, nil
+}
 
-	err = db.Ping()
-	if err != nil {
-		reg, regErr := regexp.Compile(":.*@tcp")
-		if reg == nil || regErr != nil {
-			return nil, errors.Errorf("create db connections (failed to replace password for dsn) error %v", regErr)
+// ConnectMySQL creates sql.DB used for select data
+func ConnectMySQL(cfg *mysql.Config, num int) (db *sql.DB, err error) {
+	defer func() {
+		if err != nil && db != nil {
```

Comment:
may be `if err == nil && db != nil`

Reply:
addressed in https://github.com/pingcap/tidb-tools/pull/688/commits/4f18250e1e2eb6b65015897b0a87791ee1976202